src/event/AsyncHandlerListManager.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\event;
+
+use function uasort;
+
+/**
+ * @phpstan-extends BaseHandlerListManager<AsyncEvent, AsyncRegisteredListener>
+ */
+final class AsyncHandlerListManager extends BaseHandlerListManager{
+	private static ?self $globalInstance = null;
+
+	public static function global() : self{
+		return self::$globalInstance ?? (self::$globalInstance = new self());
+	}
+
+	protected function getBaseEventClass() : string{
+		return AsyncEvent::class;
+	}
+
+	/**
+	 * @phpstan-param array<int, AsyncRegisteredListener> $listeners
+	 * @phpstan-return array<int, AsyncRegisteredListener>
+	 */
+	private static function sortSamePriorityHandlers(array $listeners) : array{
+		uasort($listeners, function(AsyncRegisteredListener $left, AsyncRegisteredListener $right) : int{
+			//While the system can handle these in any order, it's better for latency if concurrent handlers
+			//are processed together. It doesn't matter whether they are processed before or after exclusive handlers.
+			if($right->canBeCalledConcurrently()){
+				return $left->canBeCalledConcurrently() ? 0 : 1;
+			}
+			return -1;
```

Comment:
But you end up defining an order here right? Ends up `$left > $right` if `$left` is exclusive and `$right` is concurrent.

Also why do we return `-1` when both handlers are exclusive? Shouldn't it return `0` instead?

```suggestion
			//are processed together. Exclusive handlers are sorted before concurrent handlers.
			$leftConcurrent = $left->canBeCalledConcurrently();
			$rightConcurrent = $right->canBeCalledConcurrently();
			if($leftConcurrent !== $rightConcurrent){
				return $leftConcurrent <=> $rightConcurrent;
			}
			return 0;
```

Reply:
Yes, the order is intended to group concurrent handlers together, whether at the start or the end.

Good catch on the sameness. I knew there was something I didn't like about this code but I couldn't put my finger on it.

Reply:
I don't like specifying that exclusive handlers should come first without reasoning. There is no reason for them to come first. What we actually want is to aggregate concurrent handlers for Promise::all() to be efficiently used.